shared/diagram.py

```diff
@@ -45,14 +133,102 @@ def make_directories():
             else:
                 print("Successfully created the directory %s " % PATH_DIAGRAM_OUTPUT)
 
+    def group_by_group(
+        self, resources: List[Resource], initial_resource_relations: List[ResourceEdge]
+    ) -> Dict[str, List[Resource]]:
+        # Ordering Resource list to group resources into cluster
+        ordered_resources: Dict[str, List[Resource]] = dict()
+        for resource in resources:
+            if Mapsources.mapresources.get(resource.digest.type) is not None:
+                if resource.group in ordered_resources:
+                    ordered_resources[resource.group].append(resource)
+                else:
+                    ordered_resources[resource.group] = [resource]
+        return ordered_resources
+
+    def process_relationships(
+        self,
+        grouped_resources: Dict[str, List[Resource]],
+        resource_relations: List[ResourceEdge],
+    ) -> List[ResourceEdge]:
+        return resource_relations
+
     @exception
-    def generate_diagram(self, resources: List[Resource]):
-        raise NotImplementedError("Implement the diagram generation logic")
+    def generate_diagram(
+        self, resources: List[Resource], initial_resource_relations: List[ResourceEdge]
+    ):
+        ordered_resources = self.group_by_group(resources, initial_resource_relations)
+        relations = self.process_relationships(
+            ordered_resources, initial_resource_relations
+        )
+
+        with Diagram(
+            name=self.name,
+            filename=PATH_DIAGRAM_OUTPUT + self.filename,
+            direction="TB",
+            graph_attr={"nodesep": "2.0", "ranksep": "1.0", "splines": "curved"},
+        ) as d:
+            d.dot.engine = self.engine
+
+            self.draw_diagram(ordered_resources=ordered_resources, relations=relations)
+
+    def draw_diagram(self, ordered_resources, relations):
+        already_drawn_elements = {}
+
+        # Import all AWS nodes
+        for module in Mapsources.diagrams_modules:
+            exec("from diagrams.aws." + module + " import *")
+
+        nodes: Dict[ResourceDigest, any] = {}
+        # Iterate resources to draw it
+        for group_name in ordered_resources:
+            if group_name == "":
+                for resource in ordered_resources[group_name]:
+                    node = eval(Mapsources.mapresources.get(resource.digest.type))(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/gh/Cloud-Architects/cloud-discovery/pullRequest?prid=5655786)